File: content/develop/use-cases/cache-aside/dotnet/RedisCache.cs
using System.Security.Cryptography;
using StackExchange.Redis;

namespace CacheAsideDemo;

/// <summary>
/// Cache-aside helper backed by Redis hashes with TTL and Lua-backed
/// single-flight stampede protection.
/// </summary>
public class RedisCache
{
    private const string AcquireLockScript = @"
        if redis.call('SET', KEYS[1], ARGV[1], 'NX', 'PX', ARGV[2]) then
            return 1
        end
        return 0
    ";

    private const string ReleaseLockScript = @"
        if redis.call('GET', KEYS[1]) == ARGV[1] then
            return redis.call('DEL', KEYS[1])
        end
        return 0
    ";

    private readonly IDatabase _db;
    private readonly string _prefix;
    private readonly int _ttl;
    private readonly int _lockTtlMs;
    private readonly int _waitPollMs;

    private long _hits;
    private long _misses;
    private long _stampedesSuppressed;

    public RedisCache(IDatabase db, string prefix = "cache:product:", int ttl = 30,
                     int lockTtlMs = 2000, int waitPollMs = 25)
    {
        _db = db ?? throw new ArgumentNullException(nameof(db));
        if (ttl < 1) throw new ArgumentException("ttl must be at least 1 second", nameof(ttl));
        _prefix = string.IsNullOrEmpty(prefix) ? "cache:product:" : prefix;
        _ttl = ttl;
        _lockTtlMs = lockTtlMs;
        _waitPollMs = waitPollMs;
    }

    public int Ttl => _ttl;

    public sealed record Result(Dictionary<string, string>? Record, bool Hit, double RedisLatencyMs);

    /// <summary>Read through the cache, calling <paramref name="loader"/> on a miss.</summary>
    public Result Get(string entityId, Func<string, Dictionary<string, string>?> loader)
    {
        var cacheKey = CacheKey(entityId);
        var sw = System.Diagnostics.Stopwatch.StartNew();
        var entries = _db.HashGetAll(cacheKey);
        sw.Stop();
        var redisLatencyMs = sw.Elapsed.TotalMilliseconds;

        if (entries.Length > 0)
        {
            Interlocked.Increment(ref _hits);
            return new Result(ToDict(entries), Hit: true, redisLatencyMs);
        }

        Interlocked.Increment(ref _misses);
        var record = LoadWithSingleFlight(entityId, loader);
        return new Result(record, Hit: false, redisLatencyMs);
    }

    public bool Invalidate(string entityId)
    {
        return _db.KeyDelete(CacheKey(entityId));
    }

    /// <summary>
    /// Update a single field of a cached entity in place, refreshing the TTL.
    /// Only writes if the entry is already cached.
    /// </summary>
    public bool UpdateField(string entityId, string field, string value)
    {
        var cacheKey = CacheKey(entityId);
        while (true)
        {
            var tx = _db.CreateTransaction();
            tx.AddCondition(Condition.KeyExists(cacheKey));
            _ = tx.HashSetAsync(cacheKey, field, value);
            _ = tx.KeyExpireAsync(cacheKey, TimeSpan.FromSeconds(_ttl));
            if (tx.Execute())
            {
                return true;
            }
            // Condition failed (key absent) or transaction was discarded.
            if (!_db.KeyExists(cacheKey))
            {
                return false;
            }
            // Otherwise retry — the key still exists but the transaction
            // raced with another writer.
        }
    }

    public long TtlRemaining(string entityId)
    {
        var ttl = _db.KeyTimeToLive(CacheKey(entityId));
        return ttl is null ? -1 : (long) ttl.Value.TotalSeconds;
    }

    public Dictionary<string, object> Stats()
    {
        var hits = Interlocked.Read(ref _hits);
        var misses = Interlocked.Read(ref _misses);
        var stampedes = Interlocked.Read(ref _stampedesSuppressed);
        var total = hits + misses;
        var hitRatePct = total == 0 ? 0.0 : Math.Round(100.0 * hits / total, 1);
        return new Dictionary<string, object>
        {
            ["hits"] = hits,
            ["misses"] = misses,
            ["stampedes_suppressed"] = stampedes,
            ["hit_rate_pct"] = hitRatePct,
        };
    }

    public void ResetStats()
    {
        Interlocked.Exchange(ref _hits, 0);
        Interlocked.Exchange(ref _misses, 0);
        Interlocked.Exchange(ref _stampedesSuppressed, 0);
    }

    private Dictionary<string, string>? LoadWithSingleFlight(
        string entityId,
        Func<string, Dictionary<string, string>?> loader)
    {
        var cacheKey = CacheKey(entityId);
        var lockKey = LockKey(entityId);
        var token = RandomToken();

        var acquired = (long) _db.ScriptEvaluate(
            AcquireLockScript,
            new RedisKey[] { lockKey },
            new RedisValue[] { token, _lockTtlMs.ToString() });

        if (acquired == 1)
        {
            try
            {
                var record = loader(entityId);
                if (record is null) return null;
                var batch = _db.CreateBatch();
                _ = batch.KeyDeleteAsync(cacheKey);
                _ = batch.HashSetAsync(cacheKey,
                    record.Select(p => new HashEntry(p.Key, p.Value)).ToArray());
                _ = batch.KeyExpireAsync(cacheKey, TimeSpan.FromSeconds(_ttl));
                batch.Execute();
                return record;
            }
            finally
            {
                try
                {
                    _db.ScriptEvaluate(
                        ReleaseLockScript,
                        new RedisKey[] { lockKey },
                        new RedisValue[] { token });
                }
                catch
                {
                    // Lock will expire on its own.
                }
            }
        }

        Interlocked.Increment(ref _stampedesSuppressed);
        var deadline = Environment.TickCount + _lockTtlMs;
        while (Environment.TickCount < deadline)
        {
            Thread.Sleep(_waitPollMs);
            var entries = _db.HashGetAll(cacheKey);
            if (entries.Length > 0)
            {
                return ToDict(entries);
            }
        }
        return loader(entityId);
    }

    private static Dictionary<string, string> ToDict(HashEntry[] entries)
    {
        var result = new Dictionary<string, string>(entries.Length, StringComparer.Ordinal);
        foreach (var entry in entries)
        {
            result[entry.Name!] = entry.Value!;
        }
        return result;
    }

    private static string RandomToken()
    {
        Span<byte> buffer = stackalloc byte[8];
        RandomNumberGenerator.Fill(buffer);
        return Convert.ToHexString(buffer);
    }

    private string CacheKey(string id) => _prefix + id;
    private string LockKey(string id) => "lock:" + _prefix + id;
}

File: content/develop/use-cases/cache-aside/java-jedis/RedisCache.java
import java.security.SecureRandom;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Function;

import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.Transaction;
import redis.clients.jedis.exceptions.JedisException;

/**
 * Cache-aside helper backed by Redis hashes with TTL and Lua-backed
 * single-flight stampede protection.
 */
public class RedisCache {

    private static final String ACQUIRE_LOCK_SCRIPT =
            "if redis.call('SET', KEYS[1], ARGV[1], 'NX', 'PX', ARGV[2]) then\n" +
            "    return 1\n" +
            "end\n" +
            "return 0\n";

    private static final String RELEASE_LOCK_SCRIPT =
            "if redis.call('GET', KEYS[1]) == ARGV[1] then\n" +
            "    return redis.call('DEL', KEYS[1])\n" +
            "end\n" +
            "return 0\n";

    private static final SecureRandom RANDOM = new SecureRandom();

    private final JedisPool pool;
    private final String prefix;
    private final int ttl;
    private final int lockTtlMs;
    private final int waitPollMs;

    private final AtomicLong hits = new AtomicLong();
    private final AtomicLong misses = new AtomicLong();
    private final AtomicLong stampedesSuppressed = new AtomicLong();

    public RedisCache(JedisPool pool) {
        this(pool, "cache:product:", 30, 2000, 25);
    }

    public RedisCache(JedisPool pool, String prefix, int ttl, int lockTtlMs, int waitPollMs) {
        if (pool == null) {
            throw new IllegalArgumentException("pool is required");
        }
        if (ttl < 1) {
            throw new IllegalArgumentException("ttl must be at least 1 second");
        }
        this.pool = pool;
        this.prefix = (prefix == null || prefix.isEmpty()) ? "cache:product:" : prefix;
        this.ttl = ttl;
        this.lockTtlMs = lockTtlMs;
        this.waitPollMs = waitPollMs;
    }

    public int getTtl() {
        return ttl;
    }

    /** Result of a cache-aside read. */
    public static final class Result {
        public final Map<String, String> record;
        public final boolean hit;
        public final double redisLatencyMs;

        public Result(Map<String, String> record, boolean hit, double redisLatencyMs) {
            this.record = record;
            this.hit = hit;
            this.redisLatencyMs = redisLatencyMs;
        }
    }

    /** Read through the cache, calling loader on a miss. */
    public Result get(String entityId, Function<String, Map<String, String>> loader) {
        String cacheKey = cacheKey(entityId);
        Map<String, String> cached;
        long startedNs = System.nanoTime();

        try (Jedis jedis = pool.getResource()) {
            cached = jedis.hgetAll(cacheKey);
        }
        double redisLatencyMs = (System.nanoTime() - startedNs) / 1_000_000.0;

        if (cached != null && !cached.isEmpty()) {
            hits.incrementAndGet();
            return new Result(cached, true, redisLatencyMs);
        }

        misses.incrementAndGet();
        Map<String, String> record = loadWithSingleFlight(entityId, loader);
        return new Result(record, false, redisLatencyMs);
    }

    /** Delete the cached entry for entityId. */
    public boolean invalidate(String entityId) {
        try (Jedis jedis = pool.getResource()) {
            return jedis.del(cacheKey(entityId)) == 1L;
        }
    }

    /**
     * Update a single field in place if the entry is cached.
     *
     * <p>Uses WATCH/MULTI/EXEC so a concurrent invalidate or repopulate cannot
     * interleave with the field write.</p>
     */
    public boolean updateField(String entityId, String field, String value) {
        String cacheKey = cacheKey(entityId);
        while (true) {
            try (Jedis jedis = pool.getResource()) {
                jedis.watch(cacheKey);
                if (!jedis.exists(cacheKey)) {
                    jedis.unwatch();
                    return false;
                }
                Transaction tx = jedis.multi();
                tx.hset(cacheKey, field, value);
                tx.expire(cacheKey, ttl);
                List<Object> result = tx.exec();
                if (result == null) {
                    continue; // WATCH detected a change — retry.
                }
                return true;
            }
        }
    }

    /** Remaining TTL in seconds, or -2/-1 per Redis semantics. */
    public long ttlRemaining(String entityId) {
        try (Jedis jedis = pool.getResource()) {
            return jedis.ttl(cacheKey(entityId));
        }
    }

    /** Hit/miss/stampede counters with the derived hit rate. */
    public Map<String, Object> stats() {
        long h = hits.get();
        long m = misses.get();
        long s = stampedesSuppressed.get();
        long total = h + m;
        double hitRate = total == 0 ? 0.0 : ((long) (1000.0 * h / total)) / 10.0;
        Map<String, Object> stats = new HashMap<>();
        stats.put("hits", h);
        stats.put("misses", m);
        stats.put("stampedes_suppressed", s);
        stats.put("hit_rate_pct", hitRate);
        return stats;
    }

    public void resetStats() {
        hits.set(0);
        misses.set(0);
        stampedesSuppressed.set(0);
    }

    private Map<String, String> loadWithSingleFlight(
            String entityId,
            Function<String, Map<String, String>> loader) {
        String cacheKey = cacheKey(entityId);
        String lockKey = lockKey(entityId);
        String token = randomToken();

        try (Jedis jedis = pool.getResource()) {
            Object acquired = jedis.eval(
                    ACQUIRE_LOCK_SCRIPT,
                    List.of(lockKey),
                    List.of(token, Integer.toString(lockTtlMs)));
            if (asInt(acquired) == 1) {
                try {
                    Map<String, String> record = loader.apply(entityId);
                    if (record == null) {
                        return null;
                    }
                    Transaction tx = jedis.multi();
                    tx.del(cacheKey);
                    tx.hset(cacheKey, record);
                    tx.expire(cacheKey, ttl);
                    tx.exec();
                    return record;
                } finally {
                    try {
                        jedis.eval(RELEASE_LOCK_SCRIPT, List.of(lockKey), List.of(token));
                    } catch (JedisException ignored) {
                        // Lock will expire on its own.
                    }
                }
            }
        }

        stampedesSuppressed.incrementAndGet();
        long deadline = System.nanoTime() + (long) lockTtlMs * 1_000_000L;
        while (System.nanoTime() < deadline) {
            try {
                Thread.sleep(waitPollMs);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return null;
            }
            try (Jedis jedis = pool.getResource()) {
                Map<String, String> cached = jedis.hgetAll(cacheKey);
                if (cached != null && !cached.isEmpty()) {
                    return cached;
                }
            }
        }
        return loader.apply(entityId);
    }

    private static int asInt(Object value) {
        if (value instanceof Long) {
            return ((Long) value).intValue();
        }
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        return 0;
    }

    private static String randomToken() {
        byte[] bytes = new byte[8];
        RANDOM.nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }

    private String cacheKey(String id) {
        return prefix + id;
    }

    private String lockKey(String id) {
        return "lock:" + prefix + id;
    }
}
